IA halo model

chrgeorgiou opened the pull request:
A draft pull request with the implementation of the halo model for intrinsic alignments within CCL.

damonge commented:
@chrgeorgiou I see you marked this as ready for review a couple of weeks ago. Are you still looking for one? If so, can you provide an example of what this tries to achieve? I see you mentioned a cclx notebook.

chrgeorgiou commented:
Hi @damonge, indeed the PR is ready for review as soon as I fix the coverage issue (which decreased 2%). I think its best if I create a CCLX dev branch with a new notebook showcasing what this CCL development is trying to achieve. I will link this here once I have finished it.

damonge commented:
OK, great. Feel free to ping any of us for a review when you think it's ready!

chrgeorgiou commented:
The development of the branch is now completed and I am happy to have it reviewed. To make this easier, I am happy to book a few hours for answering questions about what has been developed here.

chrgeorgiou commented on doc/0000-ccl_note/main.tex:
```diff
@@ -772,6 +772,16 @@ \subsubsection{Galaxy-IA cross correlation}
+	P_{\mathrm{II}, \mathrm{h}}^{s s}(\mathbf{k}, z)&=\int \mathrm{d} M n(M) f_{s}^{2}(z) \frac{\left\langle N_{s}\left(N_{s}-1\right) | M\right\rangle}{\bar{n}_{s}^{2}(z)}\left|\hat{\gamma}_{s}^{I}(\mathbf{k} | M)\right|^{2},
+  \label{eq:IAHaloModelPowerSpectra}
+\end{align}
+where $n(M)$ is the halo mass function, $f_s(z)$ is the fraction of satellites as a function of redshift, $\langle N_s|M\rangle$ is the HOD of the satellites, $\hat{\gamma}_s^I(\mathbf{k}|M)$ is the Fourier transform of the intrinsic shear field, and $\bar{n}_s$ is the average number density of satellites.
```
Yes, good point. I added some text following the equations in Fortuna et al. 2021.

chrgeorgiou commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1553,314 @@ def _Ns(self, M, a):
+        '''
+        from scipy.integrate import simps
+        x_range = np.linspace(-1 + e, 1 - e, 1024)
+        return simps((1 - x_range ** 2) ** (a / 2.) * x_range ** b, x_range)
```
Good point here as well, I have implemented this in the code. I checked that the results from the two methods were the same.

damonge replied:
ok, one question: will b always be an integer? I'm wondering about the -1^b part

chrgeorgiou commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1553,314 @@ def _Ns(self, M, a):
+            l_sum *= np.exp(1j * 2 * phik)
+        return 2 ** l * l_sum
+
+    def gamma(self, r, r_vir):
```
Done!

chrgeorgiou commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1553,314 @@ def _Ns(self, M, a):
+        # TODO: make it possible to specify the halo density profile?
+        density_prof = HaloProfileNFW(self.cM, truncated=True,
+                                      fourier_analytic=True)
+        rho = density_prof.real(cosmo, r_use, M_use, a, mass_def)
```
I changed this as well, and tested that the results are the same.

chrgeorgiou commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1553,314 @@ def _Ns(self, M, a):
+        Adapted from pyccl.halos.profiles.HaloProfile._fftlog_wrap, but
+        modified to allow computation of the Hankel transform with variable
+        spherical bessel function index.
+        '''
```
Initially, I did not want to touch code that was already there. But I see that there should be no harm from the changes I made, so I edited the `_fftlog_wrap` directly now.

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1554,262 @@ def _Ns(self, M, a):
+                                                **kwargs)
+        self.update_precision_fftlog(padding_lo_fftlog=1E-2,
+                                     padding_hi_fftlog=1E3,
+                                     n_per_decade=512*1,
```
Do you always need this many samples, or is this only for a high-precision benchmark? Also, I guess `512*1=512` :-P

chrgeorgiou replied:
I reduced this to 100, which produces accurate enough angular power spectra. I tested this against a highly sampled Simpsons integration.

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1554,262 @@ def _Ns(self, M, a):
+    def __init__(self, c_M_relation, a1h=0.001, b=-2,
+                 lmax=6, integration_method='FFTLog',
+                 **kwargs):
+        '''
```
The docstring describing the init method usually goes after the class declaration. Check how we do it here for instance https://github.com/LSSTDESC/CCL/blob/d5408a777ab5fc4b2ae6fada2114a168b9bacee4/pyccl/halos/profiles.py#L624

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1554,262 @@ def _Ns(self, M, a):
+            c_M_relation: concentration-mass relation for the halo model.
+            a1h: Amplitude of the satellite intrinsic shear profile.
+            b: Power-law index of the satellite intrinsic shear profile.
+            If zero, the profile is assumed to be constant inside the halo.
```
see the syntax for Args here https://github.com/LSSTDESC/CCL/blob/d5408a777ab5fc4b2ae6fada2114a168b9bacee4/pyccl/halos/profiles.py#L645
My understanding is that you need to leave 4 spaces when starting a new line on the same argument. It helps with readability, but it may also be that otherwise rtd won't be able to parse them.

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1554,262 @@ def _Ns(self, M, a):
+
+        return prof
+
+    def _usat_fourier(self, cosmo, k, M, a, mass_def):
```
OK, I'm trying to understand the logic for this one. So you are overloading the `_usat_fourier` method in the HOD profile so that it now returns the satellite shear profile, right?

But then, this will go into [this line](https://github.com/LSSTDESC/CCL/blob/d5408a777ab5fc4b2ae6fada2114a168b9bacee4/pyccl/halos/profiles.py#L1507), which also contains a contribution from centrals to the total number density. What's happening with that contribution?

chrgeorgiou replied:
That is correct. The contribution from centrals is going to be neglected because of [this line](https://github.com/LSSTDESC/CCL/blob/1ce402c63ca7b01c739b32c55c1fdfce7450a922/pyccl/halos/profiles.py#L1634), where the `fc_0=0` and `ns_independent=True`.

The idea behind this is that, for the alignments between centrals we can use the NLA model. There might be a better way to do this, however, so I am happy to discuss this.

damonge reviewed (changes requested):
Second review!

One overall question: I imagine it's hard to find a benchmark for this. Can you say something about what you've done to ensure that the calculation implemented here is correct?

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1551,3 +1553,314 @@ def _Ns(self, M, a):
         return np.heaviside(M-M0, 1) * (np.fabs(M-M0) / M1)**alpha
+
+
+class SatelliteShearHOD(HaloProfileHOD):
```
ok, see my question below about centrals

chrgeorgiou commented:
> Second review!
> 
> One overall question: I imagine it's hard to find a benchmark for this. Can you say something about what you've done to ensure that the calculation implemented here is correct?

True, the benchmark is a bit tricky. I have tested against a development version of CosmoSIS, originally developed by Maria Cristina Fortuna, that the computation of $\gamma^I(k,M)$ is correct. Taking this further is difficult mostly because of the different HOD the two codes employ.

damonge commented:
@chrgeorgiou just checking that you saw my last review!

chrgeorgiou commented:
Hi @damonge, here is my reply to the last points you mentioned.

> OK, final items for review:
> 
>     1. HOD parameters: right now users could mess things up if they pass HOD parameters for centrals. 

I have changed the script to throw a warning if the user ever sets ```fc_0``` or ```fc_p``` at a value different than zero. These are forced to zero. 

>     2. Benchmarks: if you checked against a prediction by Maria Cristina Fortuna, could we add that as a benchmark?

Yes, I believe we can add this as a benchmark, let me know how I can do that. 

Note I have also added a bit of documentation based on a couple of tests I have carried out. If needed, I can add figures to support those tests or share a documented notebook that carries them out but maybe it's okay to leave things as they are.

chrgeorgiou commented:
@damonge pinging here for PR status. Changes have been completed, let me know if there is anything else needed for this PR.

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1711,3 +1730,282 @@ def _Ns(self, M, a):
+        if 'fc_0' in kwargs.keys() or 'fc_p' in kwargs.keys():
+            warnings.warn(
+                'Provided fractions for central galaxies are not zero.'
+                'Setting them to zero (fc_0=fc_p=0).',
```
I guess you either need to check that they are not zero or change the message to say that these parameters are not tweakable (I'm thinking of an overzealous user who decides to pass fc_0=0 and fc_p=0 instead of not passing anything)

damonge commented on pyccl/halos/profiles.py:
```diff
@@ -1711,3 +1730,282 @@ def _Ns(self, M, a):
+        M_use = hmc._mass
+        ngal = simps(self._Nc(M_use, a) * (1+self._Ns(M_use, a)) *
+                     hmc._massfunc.get_mass_function(cosmo, a=a, M=M_use),
+                     np.log10(M_use))
```
you can use `hmc._lmass` and avoid taking an extra logarithm.

damonge commented on benchmarks/test_haloprofile.py:
```diff
@@ -177,3 +177,28 @@ def test_weak_lensing_functions():
+                                                    1 / (1 + z_eval),
+                                                    hmd_200m)
+
+    assert np.all(np.fabs(-gamma_k_CCL / gamma_k_CosmoSIS - 1) < 0.8)
```
Humm, isn't this precision very bad? Also, what's with the preceding -1 sign in gamma_k_CCL?

damonge replied:
Can you add a comment for us to improve this benchmark in the future?

damonge reviewed (changes requested):
@chrgeorgiou very sorry for the delay. Here's a review. I mostly want to check I understand the logic behind the text.

chrgeorgiou commented:
Hi @damonge, I have pushed a new PR addressing the final comments on this. 
- On testing the new `_prefactor`, I wasn't sure what to add and I hope what I developed is sufficient. 
- On the benchmarks, the external benchmark from a development version of CosmoSIS is not accurate enough to be trusted quantitatively but serves only as a qualitative comparison. I developed code to rigorously test the accuracy of the FFTLog integration method and added the code I used for this. 

Let me know if there is anything else!
